src/Persistence/MassTransit.MongoDbIntegration/MongoDbIntegration/Outbox/MongoDbOutboxContextFactory.cs

```diff
@@ -27,21 +27,18 @@ public async Task Send<T>(ConsumeContext<T> context, OutboxConsumeOptions option
 
             FilterDefinitionBuilder<InboxState> builder = Builders<InboxState>.Filter;
             FilterDefinition<InboxState> filter = builder.Eq(x => x.MessageId, messageId) & builder.Eq(x => x.ConsumerId, options.ConsumerId);
+            var inboxStateCollection = _dbContext.GetCollection<InboxState>();
+            InboxState inboxState = null;
 
             async Task<bool> Execute()
             {
                 await _dbContext.BeginTransaction(context.CancellationToken).ConfigureAwait(false);
 
                 try
                 {
-                    MongoDbCollectionContext<InboxState> inboxStateCollection = _dbContext.GetCollection<InboxState>();
-
-                    UpdateDefinition<InboxState> update = Builders<InboxState>.Update.Set(x => x.LockToken, ObjectId.GenerateNewId());
-
-                    var inboxState = await inboxStateCollection.Lock(filter, update, context.CancellationToken).ConfigureAwait(false);
-
-                    bool continueProcessing;
+                    inboxState ??= await inboxStateCollection.Find(filter).FirstOrDefaultAsync(context.CancellationToken).ConfigureAwait(false);
```

Comment:
What if another instance of the delivery service updated the state between calls? Or if two instances try to insert the row at the same time, run the consumer, and then boom, the consumer logic runs twice. Bad.

Reply:
That will also happen if consumer 1 of the first instance inserts the inbox state and the consumer of the second instance receives it while the consume is in process in consumer 1. 
Doesn't it?

Reply:
I have changed it, so the state is always retrieved. Also the insertOne is performed before the next pipeline step is executed. This should solve the concurrent processing issue, because the mongodb transaction validates the insert for write conflicts even if the transaction is not commited.